src/compile/data/impute.ts

```diff
@@ -1,14 +1,92 @@
-import {normalizeimputeProperties} from '../../impute';
+import {isArray} from 'vega-util';
+import {Encoding} from '../../encoding';
+import {getFieldDef, isFieldDef, vgField} from '../../fielddef';
 import {ImputeSequence, ImputeTransform, isImputeSequence} from '../../transform';
-import {duplicate} from '../../util';
+import {duplicate, keys} from '../../util';
 import {VgFormulaTransform, VgImputeTransform, VgSignalRef, VgWindowTransform} from '../../vega.schema';
 import {UnitModel} from '../unit';
 import {DataFlowNode} from './dataflow';
 
-// TODO Rename to ImputeNode
-export class ImputeTransformNode extends DataFlowNode {
+export function normalizeimputeProperties(encoding: Encoding<string>): ImputeTransform | undefined {
+  const xDef = encoding.x;
+  const yDef = encoding.y;
+
+  if (isFieldDef(xDef) && isFieldDef(yDef)) {
+
+    const imputedChannel = xDef.impute ? xDef : (yDef.impute ? yDef: undefined);
+    if (imputedChannel === undefined) {
+      return undefined;
+    }
+    const keyChannel = xDef.impute ? yDef : (yDef.impute ? xDef: undefined);
+    const {method, value, frame} = imputedChannel.impute;
+    const groupbyFields = deriveEncodingGroupby(encoding);
+
+    return {
+      impute: imputedChannel.field,
+      key: keyChannel.field,
+      ...(method ? {method}:{}),
+      ...(value ? {value} : {}),
+      ...(frame ? {frame} : {}),
+      ...(groupbyFields.length ? {groupby: groupbyFields} : {} )
+    };
+  }
+  return undefined;
+}
+
+// Behavior almost identical to pathGroupingFields but without the Mark argument
+export function deriveEncodingGroupby(encoding: Encoding<string>): string[] {
```

Comment:
Why without the mark argument? 

1) This is a part of impute in encoding, which means you have the mark in your context.
2) Even if you really shouldn't have mark parameter here, copying almost identical code like this will cause maintainability issue in the future.